<a href="https://colab.research.google.com/github/arrbxr/EDA/blob/main/Data_transformation.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Data Transformation
This chepter will cover the following topics:

* Background
* Merging database-style dataframes
* Transformation techniques
* Benfits of data transformation




In [ ]:
# We will be using the following Python libraries:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

# Background
Data transformation is a set of techniques used to convert data from one format or structure to another format or structure. 



# Merging database-style dataframes

In [ ]:
# creating two dataframe
dataFrame1 =  pd.DataFrame({ 'StudentID': [1, 3, 5, 7, 9, 11, 13, 15, 17, 19, 21, 23, 25, 27, 29], 'Score' : [89, 39, 50, 97, 22, 66, 31, 51, 71, 91, 56, 32, 52, 73, 92]})
dataFrame2 =  pd.DataFrame({'StudentID': [2, 4, 6, 8, 10, 12, 14, 16, 18, 20, 22, 24, 26, 28, 30], 'Score': [98, 93, 44, 77, 69, 56, 31, 53, 78, 93, 56, 77, 33, 56, 27]})


In [ ]:
dataFrame1.head(3)

StudentID  Score
0          1     89
1          3     39
2          5     50

In [ ]:
dataFrame2.head(3)

StudentID  Score
0          2     98
1          4     93
2          6     44

In [ ]:
# # We can do that by using Pandas concat() method. 

dataframe = pd.concat([dataFrame1, dataFrame2], ignore_index=True)
dataframe

StudentID  Score
0           1     89
1           3     39
2           5     50
3           7     97
4           9     22
5          11     66
6          13     31
7          15     51
8          17     71
9          19     91
10         21     56
11         23     32
12         25     52
13         27     73
14         29     92
15          2     98
16          4     93
17          6     44
18          8     77
19         10     69
20         12     56
21         14     31
22         16     53
23         18     78
24         20     93
25         22     56
26         24     77
27         26     33
28         28     56
29         30     27

The argument `ignore_index` creates new index and its absense keeps the original indices. Note, we combined the dataframes along `axis=0`, that is to say, we **combined together along same direction.** What if we want to combine both side by side. Then we have to specify `axis = 1`. Check the output and see the difference.

In [ ]:
pd.concat([dataFrame1, dataFrame2], axis=1)

StudentID  Score  StudentID  Score
0           1     89          2     98
1           3     39          4     93
2           5     50          6     44
3           7     97          8     77
4           9     22         10     69
5          11     66         12     56
6          13     31         14     31
7          15     51         16     53
8          17     71         18     78
9          19     91         20     93
10         21     56         22     56
11         23     32         24     77
12         25     52         26     33
13         27     73         28     56
14         29     92         30     27

# Merging
In the first example, you received two files for same subject. Now, consider the use case where you are teaching two courses. So, you will get two dataframes from each sections: two for Software engieering course and another two for Introduction to Machine learning course. Check the figure given below:

In [ ]:
df1SE =  pd.DataFrame({ 'StudentID': [9, 11, 13, 15, 17, 19, 21, 23, 25, 27, 29], 'ScoreSE' : [22, 66, 31, 51, 71, 91, 56, 32, 52, 73, 92]})
df2SE =  pd.DataFrame({'StudentID': [2, 4, 6, 8, 10, 12, 14, 16, 18, 20, 22, 24, 26, 28, 30], 'ScoreSE': [98, 93, 44, 77, 69, 56, 31, 53, 78, 93, 56, 77, 33, 56, 27]})

df1ML =  pd.DataFrame({ 'StudentID': [1, 3, 5, 7, 9, 11, 13, 15, 17, 19, 21, 23, 25, 27, 29], 'ScoreML' : [39, 49, 55, 77, 52, 86, 41, 77, 73, 51, 86, 82, 92, 23, 49]})
df2ML =  pd.DataFrame({'StudentID': [2, 4, 6, 8, 10, 12, 14, 16, 18, 20], 'ScoreML': [93, 44, 78, 97, 87, 89, 39, 43, 88, 78]})

In [ ]:
df1SE.head(3)

StudentID  ScoreSE
0          9       22
1         11       66
2         13       31

In [ ]:
df2ML.head(2)

StudentID  ScoreML
0          2       93
1          4       44

In [ ]:
dfse = pd.concat([df1SE, df2SE], ignore_index=True)
dfml = pd.concat([df1ML, df2ML], ignore_index=True)

df = dfse.merge(dfml, how='left')
df

StudentID  ScoreSE  ScoreML
0           9       22     52.0
1          11       66     86.0
2          13       31     41.0
3          15       51     77.0
4          17       71     73.0
5          19       91     51.0
6          21       56     86.0
7          23       32     82.0
8          25       52     92.0
9          27       73     23.0
10         29       92     49.0
11          2       98     93.0
12          4       93     44.0
13          6       44     78.0
14          8       77     97.0
15         10       69     87.0
16         12       56     89.0
17         14       31     39.0
18         16       53     43.0
19         18       78     88.0
20         20       93     78.0
21         22       56      NaN
22         24       77      NaN
23         26       33      NaN
24         28       56      NaN
25         30       27      NaN

In [ ]:
dfse = pd.concat([df1SE, df2SE], ignore_index=True)
dfml = pd.concat([df1ML, df2ML], ignore_index=True)

df = dfse.merge(dfml, how='right')
df

StudentID  ScoreSE  ScoreML
0           1      NaN       39
1           3      NaN       49
2           5      NaN       55
3           7      NaN       77
4           9     22.0       52
5          11     66.0       86
6          13     31.0       41
7          15     51.0       77
8          17     71.0       73
9          19     91.0       51
10         21     56.0       86
11         23     32.0       82
12         25     52.0       92
13         27     73.0       23
14         29     92.0       49
15          2     98.0       93
16          4     93.0       44
17          6     44.0       78
18          8     77.0       97
19         10     69.0       87
20         12     56.0       89
21         14     31.0       39
22         16     53.0       43
23         18     78.0       88
24         20     93.0       78

In [ ]:
dfse = pd.concat([df1SE, df2SE], ignore_index=True)
dfml = pd.concat([df1ML, df2ML], ignore_index=True)

df = dfse.merge(dfml, how='inner')
df

StudentID  ScoreSE  ScoreML
0           9       22       52
1          11       66       86
2          13       31       41
3          15       51       77
4          17       71       73
5          19       91       51
6          21       56       86
7          23       32       82
8          25       52       92
9          27       73       23
10         29       92       49
11          2       98       93
12          4       93       44
13          6       44       78
14          8       77       97
15         10       69       87
16         12       56       89
17         14       31       39
18         16       53       43
19         18       78       88
20         20       93       78

# Merging on index


In [ ]:
left1 = pd.DataFrame({'key': ['apple','ball','apple', 'apple',
'ball', 'cat'], 'value': range(6)})
right1 = pd.DataFrame({'group_val': [33.4, 5]}, index=['apple',
'ball'])

In [ ]:
left1

key  value
0  apple      0
1   ball      1
2  apple      2
3  apple      3
4   ball      4
5    cat      5

In [ ]:
right1

group_val
apple       33.4
ball         5.0

In [ ]:
df = pd.merge(left1, right1, left_on='key', right_index=True)
df

key  value  group_val
0  apple      0       33.4
2  apple      2       33.4
3  apple      3       33.4
1   ball      1        5.0
4   ball      4        5.0

In [ ]:
df = pd.merge(left1, right1, left_on='key', right_index=True, how='outer')
df

key  value  group_val
0  apple      0       33.4
2  apple      2       33.4
3  apple      3       33.4
1   ball      1        5.0
4   ball      4        5.0
5    cat      5        NaN

# Reshaping and pivoting
During EDA, we often need to rearrange data in a dataframe in some consistent manner.
This can be done with hierarchical indexing using two actions:

* **Stacking** `Stack` rotates from any particular column in the data to the rows
* **Unstacking** `Unstack` rotates from the rows into column

In [ ]:
data = np.arange(15).reshape((3, 5))
index = ['Rainfall', 'Humidity', 'Wind']
dframe1 = pd.DataFrame(data, index=index, columns=['Bergen', 'Oslo', 'Trondheim', 'Stavenger', 'Kristiansand'])

dframe1

Bergen  Oslo  Trondheim  Stavenger  Kristiansand
Rainfall       0     1          2          3             4
Humidity       5     6          7          8             9
Wind          10    11         12         13            14

Now, using the `stack()` method on the preceding dframe1, we can pivot the
columns into rows to produce a series:

In [ ]:
stacked = dframe1.stack()
stacked

Rainfall  Bergen           0
          Oslo             1
          Trondheim        2
          Stavenger        3
          Kristiansand     4
Humidity  Bergen           5
          Oslo             6
          Trondheim        7
          Stavenger        8
          Kristiansand     9
Wind      Bergen          10
          Oslo            11
          Trondheim       12
          Stavenger       13
          Kristiansand    14
dtype: int64

In [ ]:
stacked.unstack()

Bergen  Oslo  Trondheim  Stavenger  Kristiansand
Rainfall       0     1          2          3             4
Humidity       5     6          7          8             9
Wind          10    11         12         13            14

In [ ]:
series1 = pd.Series([000, 111, 222, 333], index=['zeros','ones','twos', 'threes'])
series2 = pd.Series([444, 555, 666], index=['fours', 'fives','sixes'])

In [ ]:
frame2 = pd.concat([series1, series2], keys=['Number1', 'Number2'])
frame2.unstack()

fives  fours   ones  sixes  threes   twos  zeros
Number1    NaN    NaN  111.0    NaN   333.0  222.0    0.0
Number2  555.0  444.0    NaN  666.0     NaN    NaN    NaN

# Performing data deduplication

In [ ]:
frame3 = pd.DataFrame({'column 1': ['Looping'] * 3 + ['Functions']
* 4, 'column 2': [10, 10, 22, 23, 23, 24, 24]})
frame3

column 1  column 2
0    Looping        10
1    Looping        10
2    Looping        22
3  Functions        23
4  Functions        23
5  Functions        24
6  Functions        24

The `pandas` dataframe comes with a `duplicated()` method that returns a
Boolean series stating which of the rows are duplicates

In [ ]:
frame3.duplicated()

0    False
1     True
2    False
3    False
4     True
5    False
6     True
dtype: bool

In [ ]:
frame4 = frame3.drop_duplicates()
frame4

column 1  column 2
0    Looping        10
2    Looping        22
3  Functions        23
5  Functions        24

Let's add a new column and try to find duplicated items based on the second
column

In [ ]:
frame3['column 3'] = range(7)
frame5 = frame3.drop_duplicates(['column 2'])
frame5

column 1  column 2  column 3
0    Looping        10         0
2    Looping        22         2
3  Functions        23         3
5  Functions        24         5

# Replacing values


In [ ]:
import numpy as np
replaceframe = pd.DataFrame({'column 1': [200., 3000., -786., 3000., 234., 444., -786., 332., 3332.],
                             'column 2': range(9)})

replaceframe.replace(to_replace = -786, value=np.nan)

column 1  column 2
0     200.0         0
1    3000.0         1
2       NaN         2
3    3000.0         3
4     234.0         4
5     444.0         5
6       NaN         6
7     332.0         7
8    3332.0         8

In [ ]:
replaceFrame = pd.DataFrame({'column 1': [200., 3000., -786.,
3000., 234., 444., -786., 332., 3332., 0 ], 'column 2': range(10)})
replaceFrame.replace(to_replace =[-786, 0], value= [np.nan, 2])


column 1  column 2
0     200.0         2
1    3000.0         1
2       NaN         2
3    3000.0         3
4     234.0         4
5     444.0         5
6       NaN         6
7     332.0         7
8    3332.0         8
9       2.0         9

# Handling missing data

In [ ]:
# Let's assume we have a dataframe as shown here:
data = np.arange(15, 30).reshape(5, 3)
dfx = pd.DataFrame(data, index=['apple', 'banana', 'kiwi', 'grapes', 'mango'], columns=['store1', 'store2', 'stoore3'])

dfx

store1  store2  stoore3
apple       15      16       17
banana      18      19       20
kiwi        21      22       23
grapes      24      25       26
mango       27      28       29

In [ ]:
# Let's add some missing values to our dataframe:
dfx['store4'] = np.nan
dfx.loc['watermelone'] = np.arange(15, 19)
dfx.loc['oranges'] = np.nan
dfx['store5'] = np.nan
dfx['store4']['apple'] = 20

dfx

store1  store2  stoore3  store4  store5
apple          15.0    16.0     17.0    20.0     NaN
banana         18.0    19.0     20.0     NaN     NaN
kiwi           21.0    22.0     23.0     NaN     NaN
grapes         24.0    25.0     26.0     NaN     NaN
mango          27.0    28.0     29.0     NaN     NaN
watermelone    15.0    16.0     17.0    18.0     NaN
oranges         NaN     NaN      NaN     NaN     NaN

In [ ]:
# Check null values
dfx.isnull()

store1  store2  stoore3  store4  store5
apple         False   False    False   False    True
banana        False   False    False    True    True
kiwi          False   False    False    True    True
grapes        False   False    False    True    True
mango         False   False    False    True    True
watermelone   False   False    False   False    True
oranges        True    True     True    True    True

In [ ]:
dfx.isnull().sum()

store1     1
store2     1
stoore3    1
store4     5
store5     7
dtype: int64

In [ ]:
dfx.notnull()

store1  store2  stoore3  store4  store5
apple          True    True     True    True   False
banana         True    True     True   False   False
kiwi           True    True     True   False   False
grapes         True    True     True   False   False
mango          True    True     True   False   False
watermelone    True    True     True    True   False
oranges       False   False    False   False   False

In [ ]:
dfx.isnull().sum().sum()

15

In [ ]:
dfx.count()

store1     6
store2     6
stoore3    6
store4     2
store5     0
dtype: int64

# Dropping missing values
One of the ways to handle missing values is to simply remove them from our dataset. We
have seen that we can use the isnull() and notnull() functions from the pandas library
to determine null values

In [ ]:
dfx.store4[dfx.store4.notnull()]

apple          20.0
watermelone    18.0
Name: store4, dtype: float64

The output shows that store4 only reported two items of data. Now, we can use the
dropna() method to remove the rows

In [ ]:
dfx['store4'].dropna()

apple          20.0
watermelone    18.0
Name: store4, dtype: float64

In [ ]:
dfx.dropna()

Empty DataFrame
Columns: [store1, store2, stoore3, store4, store5]
Index: []

# Dropping by rows

In [ ]:
dfx.dropna(how='all')

store1  store2  stoore3  store4  store5
apple          15.0    16.0     17.0    20.0     NaN
banana         18.0    19.0     20.0     NaN     NaN
kiwi           21.0    22.0     23.0     NaN     NaN
grapes         24.0    25.0     26.0     NaN     NaN
mango          27.0    28.0     29.0     NaN     NaN
watermelone    15.0    16.0     17.0    18.0     NaN

# Dropping by columns

In [ ]:
dfx.dropna(how='all', axis=1)

store1  store2  stoore3  store4
apple          15.0    16.0     17.0    20.0
banana         18.0    19.0     20.0     NaN
kiwi           21.0    22.0     23.0     NaN
grapes         24.0    25.0     26.0     NaN
mango          27.0    28.0     29.0     NaN
watermelone    15.0    16.0     17.0    18.0
oranges         NaN     NaN      NaN     NaN

Note that store5 is dropped from the dataframe. By passing in axis=1, we are instructing
pandas to drop columns if all the values in the column are NaN. Furthermore, we can also
pass another argument, `thresh`, to specify a minimum number of NaNs that must exist
before the column should be dropped:

In [ ]:
dfx.dropna(thresh=5, axis=1)

store1  store2  stoore3
apple          15.0    16.0     17.0
banana         18.0    19.0     20.0
kiwi           21.0    22.0     23.0
grapes         24.0    25.0     26.0
mango          27.0    28.0     29.0
watermelone    15.0    16.0     17.0
oranges         NaN     NaN      NaN

# Mathematical operations with NaN
The `pandas` and `numpy` libraries handle NaN values differently for mathematical
operations.

In [ ]:
ar1 = np.array([100, 200, np.nan, 300])
sr1 = pd.Series(ar1)

In [ ]:
ar1.mean(), sr1.mean()

(nan, 200.0)

In [ ]:
ar1

array([100., 200.,  nan, 300.])

In [ ]:
sr1

0    100.0
1    200.0
2      NaN
3    300.0
dtype: float64

# Note the following things

* When a NumPy function encounters `NaN` values, it returns `NaN`.
* Pandas, on the other hand, ignores the `NaN` values and moves ahead with
processing. When performing the sum operation, `NaN` is treated as `0`. If all the
values are `NaN`, the result is also `NaN`.

In [ ]:
# Let's compute the total quantity of fruits sold by store4
sr2 = dfx.store4
sr2.sum()

38.0

Note that store4 has five NaN values. However, during the summing process, these 
values are treated as 0 and the result is 38.0. 

In [ ]:
# Note that NaNs are treated as 0s. It is the same for cumulative summing:
sr2.cumsum()

apple          20.0
banana          NaN
kiwi            NaN
grapes          NaN
mango           NaN
watermelone    38.0
oranges         NaN
Name: store4, dtype: float64

# Filling missing values
We can use the `fillna()` method to replace `NaN` values with any particular values.

In [ ]:
fildf = dfx.fillna(0)

In [ ]:
fildf

store1  store2  stoore3  store4  store5
apple          15.0    16.0     17.0    20.0     0.0
banana         18.0    19.0     20.0     0.0     0.0
kiwi           21.0    22.0     23.0     0.0     0.0
grapes         24.0    25.0     26.0     0.0     0.0
mango          27.0    28.0     29.0     0.0     0.0
watermelone    15.0    16.0     17.0    18.0     0.0
oranges         0.0     0.0      0.0     0.0     0.0

Note that in the preceding dataframe, all the `NaN` values are replaced by `0`. Replacing the
values with `0` will **affect several statistics including mean, sum, and median**.


In [ ]:
dfx.mean()

store1     20.0
store2     21.0
stoore3    22.0
store4     19.0
store5      NaN
dtype: float64

Now, let's compute the mean from the filled dataframe with the following command:

In [ ]:
fildf.mean()

store1     17.142857
store2     18.000000
stoore3    18.857143
store4      5.428571
store5      0.000000
dtype: float64

Note that there are slightly different values. Hence, filling with 0 might not be the optimal
solution. 

# Backward and forward filling
`NaN` values can be filled based on the last known values. To understand this, let's consider
taking our store dataframe as an example.


In [ ]:
# We want to fill store4 using the forward-filling technique:
dfx.store4.fillna(method='ffill')

apple          20.0
banana         20.0
kiwi           20.0
grapes         20.0
mango          20.0
watermelone    18.0
oranges        18.0
Name: store4, dtype: float64

Here, from the forward-filling technique, the last known value is 20 and hence the rest of
the NaN values are replaced by it.

The direction of the fill can be changed by changing `method='bfill'`. Check the following
example:

In [ ]:
dfx['store4'].fillna(method='bfill')

apple          20.0
banana         18.0
kiwi           18.0
grapes         18.0
mango          18.0
watermelone    18.0
oranges         NaN
Name: store4, dtype: float64

# Interpolating missing values
The pandas library provides the `interpolate()` function both for the series and the
dataframe. By default, it performs a linear interpolation of our missing values. Check the
following example:


In [ ]:
ser3 = pd.Series([100, np.nan, np.nan, np.nan, 292])
ser3.interpolate()

0    100.0
1    148.0
2    196.0
3    244.0
4    292.0
dtype: float64

In [ ]:
ser3 = pd.Series([100, np.nan, 300, np.nan, np.nan, 500])
ser3.interpolate()


0    100.000000
1    200.000000
2    300.000000
3    366.666667
4    433.333333
5    500.000000
dtype: float64

Are you wondering how these values are calculated? Well, it is done by taking the first
value before and after any sequence of the NaN values. In the preceding series, ser3, the
first and the last values are 100 and 292 respectively. Hence, **it calculates the next value as
`(292-100)/(5-1) = 48`. So, the next value after 100 is `100 + 48 = 148`.**

### Note:-
**We can perform more complex interpolation techniques, especially with
time series data. An example of this interpolation is shown in the
notebook provided with this chapter.** 

# Renaming axis indexes

In [ ]:
# Say you want to transform the index terms to capital letters:
dframe1

Bergen  Oslo  Trondheim  Stavenger  Kristiansand
Rainfall       0     1          2          3             4
Humidity       5     6          7          8             9
Wind          10    11         12         13            14

In [ ]:
dframe1.index = dframe1.index.map(str.upper)
dframe1

Bergen  Oslo  Trondheim  Stavenger  Kristiansand
RAINFALL       0     1          2          3             4
HUMIDITY       5     6          7          8             9
WIND          10    11         12         13            14

If we want to create a transformed version of
the dataframe, then we can use the `rename()` method.

This method is handy when we do
not want to modify the original data. Check the following example:


In [ ]:
dframe1.rename(index=str.title, columns=str.upper)


BERGEN  OSLO  TRONDHEIM  STAVENGER  KRISTIANSAND
Rainfall       0     1          2          3             4
Humidity       5     6          7          8             9
Wind          10    11         12         13            14

# Discretization and binning
Often when working with continuous datasets, we need to convert them into discrete or 
interval forms. Each interval is referred to as a bin, and hence the name binning comes into
play:


In [ ]:
# Let's say we have data on the heights of a group of students as follows:
height = [120, 122, 125, 127, 121, 123, 137, 131, 161, 145, 141,132]

And we want to convert that dataset into intervals of 118 to 125, 126 to 135, 136 to 160, and finally 160 and higher.

To convert the preceding dataset into intervals, we can use the `cut()` method
provided by the pandas library:

In [ ]:
bins = [118, 125, 135, 160, 200]
category = pd.cut(height, bins)

category

[(118, 125], (118, 125], (118, 125], (125, 135], (118, 125], ..., (125, 135], (160, 200], (135, 160], (135, 160], (125, 135]]
Length: 12
Categories (4, interval[int64]): [(118, 125] < (125, 135] < (135, 160] < (160, 200]]

If you look closely at the output, you'll see that there are mathematical notations for intervals. Do you recall what these parentheses mean from your elementary mathematics class? If not, here is a quick recap:

* A parenthesis indicates that the side is open.
* A square bracket means that it is closed or inclusive. 

From the preceding code block, `(118, 125]` means the left-hand side is open
and the right-hand side is closed. This is mathematically denoted as follows:

<h1><center>(a, b] = {x|a &lt; x &le; b}</center></h1>

Hence, 118 is not included, but anything greater than 118 is included, while 125
is included in the interval.


We can set a `right=False` argument to change the form of interval

In [ ]:
category2 = pd.cut(height, [118, 126, 136, 161, 200], right=False)
category2

[[118, 126), [118, 126), [118, 126), [126, 136), [118, 126), ..., [126, 136), [161, 200), [136, 161), [136, 161), [126, 136)]
Length: 12
Categories (4, interval[int64]): [[118, 126) < [126, 136) < [136, 161) < [161, 200)]

Note that the output form of closeness has been changed. Now, the results are in
the form of **right-closed, left-open**


We can check the number of values in each bin by using the
`pd.value_counts()` method:

In [ ]:
pd.value_counts(category)

(118, 125]    5
(135, 160]    3
(125, 135]    3
(160, 200]    1
dtype: int64

The output shows that there are five values in the interval [118-125)

In [ ]:
pd.value_counts(category2)

[118, 126)    5
[136, 161)    3
[126, 136)    3
[161, 200)    1
dtype: int64

#### We can also indicate the bin names by passing a list of labels:

In [ ]:
bin_name = ['Short Height', 'Average Height', 'Good Height', 'Taller']
pd.cut(height, bins, labels=bin_name)

['Short Height', 'Short Height', 'Short Height', 'Average Height', 'Short Height', ..., 'Average Height', 'Taller', 'Good Height', 'Good Height', 'Average Height']
Length: 12
Categories (4, object): ['Short Height' < 'Average Height' < 'Good Height' < 'Taller']

Note that we have passed at least two arguments, the data that needs to be
discretized and the required number of bins. 

Furthermore, we have used a `right=False` argument to change the form of interval.

 Now, it is essential to note that if we pass just an integer for our bins, it will
compute equal-length bins based on the minimum and maximum values in the
data. Okay, let's verify what we mentioned here:

In [ ]:
import numpy as np
pd.cut(np.random.rand(40), 5, precision=2)

[(0.24, 0.43], (0.81, 1.0], (0.43, 0.62], (0.62, 0.81], (0.048, 0.24], ..., (0.81, 1.0], (0.81, 1.0], (0.81, 1.0], (0.048, 0.24], (0.43, 0.62]]
Length: 40
Categories (5, interval[float64]): [(0.048, 0.24] < (0.24, 0.43] < (0.43, 0.62] < (0.62, 0.81] <
                                    (0.81, 1.0]]

Pandas provides a qcut method that forms the bins based on sample quantiles. Let's check
this with an example

In [ ]:
randomNumbers = np.random.rand(2000)
category3 = pd.qcut(randomNumbers, 4) # cut into quartiles
category3

[(0.744, 1.0], (0.744, 1.0], (0.493, 0.744], (0.744, 1.0], (0.493, 0.744], ..., (0.0004499999999999999, 0.24], (0.744, 1.0], (0.493, 0.744], (0.24, 0.493], (0.0004499999999999999, 0.24]]
Length: 2000
Categories (4, interval[float64]): [(0.0004499999999999999, 0.24] < (0.24, 0.493] < (0.493, 0.744] < (0.744, 1.0]]

Note that based on the number of bins, which we set to 4, it converted our data into four
different categories. If we count the number of values in each category, we should get
equal-sized bins as per our definition. Let's verify that with the following command:

In [ ]:
pd.value_counts(category3)


(0.744, 1.0]                     500
(0.493, 0.744]                   500
(0.24, 0.493]                    500
(0.0004499999999999999, 0.24]    500
dtype: int64

Our claim is hence verified. Each category contains an equal size of 500 values. Note that,
similar to cut, we can also pass our own bins:

In [ ]:
pd.qcut(randomNumbers, [0, 0.3, 0.5, 0.7, 1.0])

[(0.702, 1.0], (0.702, 1.0], (0.702, 1.0], (0.702, 1.0], (0.493, 0.702], ..., (0.0004499999999999999, 0.291], (0.702, 1.0], (0.702, 1.0], (0.291, 0.493], (0.0004499999999999999, 0.291]]
Length: 2000
Categories (4, interval[float64]): [(0.0004499999999999999, 0.291] < (0.291, 0.493] < (0.493, 0.702] < (0.702, 1.0]]

# Outlier detection and filtering

Outliers are data points that diverge from other observations for several reasons. During
the EDA phase, one of our common tasks is to detect and filter these outliers. The main
reason for this detection and filtering of outliers is that the presence of such outliers can
cause serious issues in statistical analysis. In this section, we are going to perform simple
outlier detection and filtering. Let's get started

In [ ]:
# Load the dataset that is available from the GitHub link as follows:
df = pd.read_csv('https://raw.githubusercontent.com/arrbxr/Hands-on-Exploratory-Data-Analysis-with-Python/master/Chapter%204/sales.csv')
df.head(8)

Account           Company  Order  ... Quantity UnitPrice  transactionComplete
0  123456779         Kulas Inc  99985  ...     5148       545                False
1  123456784            GitHub  99986  ...     3262       383                False
2  123456782         Kulas Inc  99990  ...     9119       407                 True
3  123456783         My SQ Man  99999  ...     3097       615                False
4  123456787         ABC Dogma  99996  ...     3356        91                 True
5  123456778  Super Sexy Dingo  99996  ...     2474       136                 True
6  123456783         ABC Dogma  99981  ...     4081       195                False
7  123456785         ABC Dogma  99998  ...     6576       603                False

[8 rows x 9 columns]

Now, suppose we want to calculate the total price based on the quantity sold and
the unit price.

We can simply add a new column, as shown here:

In [ ]:
df['Total_price'] = df["UnitPrice"] * df["Quantity"]
df.head()

Account    Company  Order  ... UnitPrice transactionComplete  Total_price
0  123456779  Kulas Inc  99985  ...       545               False      2805660
1  123456784     GitHub  99986  ...       383               False      1249346
2  123456782  Kulas Inc  99990  ...       407                True      3711433
3  123456783  My SQ Man  99999  ...       615               False      1904655
4  123456787  ABC Dogma  99996  ...        91                True       305396

[5 rows x 10 columns]

Now, let's answer some questions based on the preceding table


Let's find the transaction that exceeded 3,000,000:

In [ ]:
TotalTransaction = df["Total_price"] 
TotalTransaction[np.abs(TotalTransaction) > 3000000]

2       3711433
7       3965328
13      4758900
15      5189372
17      3989325
         ...   
9977    3475824
9984    5251134
9987    5670420
9991    5735513
9996    3018490
Name: Total_price, Length: 2094, dtype: int64

Note that, in the preceding example, we have assumed that any price greater than 3,000,000
is an outlier. 

Display all the columns and rows from the preceding table if TotalPrice is greater
than 6741112, as follows:


In [ ]:
df[np.abs(TotalTransaction) > 6741112]

Account    Company  Order  ... UnitPrice transactionComplete  Total_price
818   123456781  Gen Power  99991  ...       696               False      6746328
1402  123456778   Will LLC  99985  ...       695                True      6841580
2242  123456770    Name IT  99997  ...       692               False      6784368
2876  123456772  Gen Power  99992  ...       679               False      6745865
3210  123456782  Loolo INC  99991  ...       692               False      6841112
3629  123456779  My SQ Man  99980  ...       700               False      6785800
7674  123456781  Loolo INC  99989  ...       691               False      6828462
8645  123456789  Gen Power  99996  ...       699               False      6809658
8684  123456785  Gen Power  99989  ...       694               False      6804670

[9 rows x 10 columns]

Note that in the output, all the TotalPrice values are greater than 6741112. We can use
any sort of conditions, either row-wise or column-wise, to detect and filter outliers.

# Permutation and random sampling
Well, now we have some more mathematical terms to learn: permutation and random
sampling. Let's examine how we can perform permutation and random sampling using the
pandas library

* With NumPy's numpy.random.permutation() function, we can randomly
select or permute a series of rows in a dataframe. Let's understand this with an
example:

In [ ]:
dat = np.arange(80).reshape(10, 8)
df = pd.DataFrame(dat)
df

0   1   2   3   4   5   6   7
0   0   1   2   3   4   5   6   7
1   8   9  10  11  12  13  14  15
2  16  17  18  19  20  21  22  23
3  24  25  26  27  28  29  30  31
4  32  33  34  35  36  37  38  39
5  40  41  42  43  44  45  46  47
6  48  49  50  51  52  53  54  55
7  56  57  58  59  60  61  62  63
8  64  65  66  67  68  69  70  71
9  72  73  74  75  76  77  78  79

Next, we call the `np.random.permutation()` method. This method takes an
argument – the length of the axis we require to be permuted – and gives an array
of integers indicating the new ordering:

In [ ]:
sampler = np.random.permutation(10)
sampler

array([5, 8, 3, 0, 2, 1, 9, 7, 6, 4])

The preceding output array is used in ix-based indexing for the `take()` function
from the `pandas library`. Check the following example for clarification

In [ ]:
df.take(sampler)

0   1   2   3   4   5   6   7
5  40  41  42  43  44  45  46  47
8  64  65  66  67  68  69  70  71
3  24  25  26  27  28  29  30  31
0   0   1   2   3   4   5   6   7
2  16  17  18  19  20  21  22  23
1   8   9  10  11  12  13  14  15
9  72  73  74  75  76  77  78  79
7  56  57  58  59  60  61  62  63
6  48  49  50  51  52  53  54  55
4  32  33  34  35  36  37  38  39

# Random sampling without replacemet
To compute random sampling without replacement, follow these steps:

* To perform random sampling without replacement, we first create
a `permutation` array.
* Next, we slice off the first n elements of the array where n is the desired size of the subset you want to sample.
* Then we use the `df.take()` method to obtain actual samples:


In [ ]:
df.take(np.random.permutation(len(df))[:3])

0   1   2   3   4   5   6   7
6  48  49  50  51  52  53  54  55
9  72  73  74  75  76  77  78  79
7  56  57  58  59  60  61  62  63

# Computing indicators/dummy variables
Often, we need to convert a categorical variable into some dummy matrix. Especially for
statistical modeling or machine learning model development, it is essential to create
dummy variables. Let's get started



In [ ]:
df = pd.DataFrame({'gender': ['female', 'female', 'male', 'unknown', 'male', 'female'], 'votes': range(6, 12, 1)})

df

gender  votes
0   female      6
1   female      7
2     male      8
3  unknown      9
4     male     10
5   female     11

So far, nothing too complicated. Sometimes, however, we need to encode these
values in a matrix form with 1 and 0 values

We can do that using the `pd.get_dummies()` function:

In [ ]:
pd.get_dummies(df['gender'])

female  male  unknown
0       1     0        0
1       1     0        0
2       0     1        0
3       0     0        1
4       0     1        0
5       1     0        0

Note the pattern. There are five values in the original dataframe with three
unique values of `male`, `female`, and `unknown`. Each unique value is transformed
into a column and each original value into a row. For example, in the original
dataframe, the first value is `female`, hence it is added as a row with 1 in the
`female` value and the rest of them are 0 values, and so on.

Sometimes, we want to add a prefix to the columns. We can do that by adding
the prefix argument, as shown here:

In [ ]:
dummies = pd.get_dummies(df['gender'], prefix='gender')
dummies

gender_female  gender_male  gender_unknown
0              1            0               0
1              1            0               0
2              0            1               0
3              0            0               1
4              0            1               0
5              1            0               0